// rc-util/lib/PortalWrapper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactLifecyclesCompat = require('react-lifecycles-compat');

var _ContainerRender = require('./ContainerRender');

var _ContainerRender2 = _interopRequireDefault(_ContainerRender);

var _Portal = require('./Portal');

var _Portal2 = _interopRequireDefault(_Portal);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var openCount = 0;
var windowIsUndefined = !(typeof window !== 'undefined' && window.document && window.document.createElement);

var IS_REACT_16 = 'createPortal' in _reactDom2['default'];

var PortalWrapper = function (_React$Component) {
  (0, _inherits3['default'])(PortalWrapper, _React$Component);

  function PortalWrapper(props) {
    (0, _classCallCheck3['default'])(this, PortalWrapper);

    var _this = (0, _possibleConstructorReturn3['default'])(this, (PortalWrapper.__proto__ || Object.getPrototypeOf(PortalWrapper)).call(this, props));

    _initialiseProps.call(_this);

    var visible = props.visible;

    openCount = visible ? openCount + 1 : openCount;
    _this.state = {
      _self: _this
    };
    return _this;
  }

  (0, _createClass3['default'])(PortalWrapper, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.setWrapperClassName();
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      var visible = this.props.visible;
      // 离开时不会 render， 导到离开时数值不变，改用 func 。。

      openCount = visible && openCount ? openCount - 1 : openCount;
      this.removeCurrentContainer(visible);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          forceRender = _props.forceRender,
          visible = _props.visible;

      var portal = null;
      var childProps = {
        getOpenCount: function getOpenCount() {
          return openCount;
        },
        getContainer: this.getContainer
      };
      // suppport react15
      if (!IS_REACT_16) {
        return _react2['default'].createElement(
          _ContainerRender2['default'],
          {
            parent: this,
            visible: visible,
            autoDestroy: false,
            getComponent: function getComponent() {
              var extra = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
              return children((0, _extends3['default'])({}, extra, childProps, {
                ref: _this2.savePortal
              }));
            },
            getContainer: this.getContainer,
            forceRender: forceRender
          },
          function (_ref) {
            var renderComponent = _ref.renderComponent,
                removeContainer = _ref.removeContainer;

            _this2.renderComponent = renderComponent;
            _this2.removeContainer = removeContainer;
            return null;
          }
        );
      }
      if (forceRender || visible || this._component) {
        portal = _react2['default'].createElement(
          _Portal2['default'],
          { getContainer: this.getContainer, ref: this.savePortal },
          children(childProps)
        );
      }
      return portal;
    }
  }], [{
    key: 'getDerivedStateFromProps',
    value: function getDerivedStateFromProps(props, _ref2) {
      var prevProps = _ref2.prevProps,
          _self = _ref2._self;
      var visible = props.visible,
          getContainer = props.getContainer;

      if (prevProps) {
        var prevVisible = prevProps.visible,
            prevGetContainer = prevProps.getContainer;

        if (visible !== prevVisible) {
          openCount = visible && !prevVisible ? openCount + 1 : openCount - 1;
        }
        if (getContainer !== prevGetContainer) {
          _self.removeCurrentContainer(false);
        }
      }
      return {
        prevProps: props
      };
    }
  }]);
  return PortalWrapper;
}(_react2['default'].Component);

PortalWrapper.propTypes = {
  wrapperClassName: _propTypes2['default'].string,
  forceRender: _propTypes2['default'].bool,
  getContainer: _propTypes2['default'].any,
  children: _propTypes2['default'].func,
  visible: _propTypes2['default'].bool
};

var _initialiseProps = function _initialiseProps() {
  var _this3 = this;

  this.getParent = function () {
    var getContainer = _this3.props.getContainer;

    if (getContainer) {
      if (typeof getContainer === 'string') {
        return document.querySelectorAll(getContainer)[0];
      }
      if (typeof getContainer === 'function') {
        return getContainer();
      }
      if ((typeof getContainer === 'undefined' ? 'undefined' : (0, _typeof3['default'])(getContainer)) === 'object' && getContainer instanceof window.HTMLElement) {
        return getContainer;
      }
    }
    return document.body;
  };

  this.getContainer = function () {
    if (windowIsUndefined) {
      return null;
    }
    if (!_this3.container) {
      _this3.container = document.createElement('div');
      var parent = _this3.getParent();
      parent.appendChild(_this3.container);
    }
    _this3.setWrapperClassName();
    return _this3.container;
  };

  this.setWrapperClassName = function () {
    var wrapperClassName = _this3.props.wrapperClassName;

    if (_this3.container && wrapperClassName && wrapperClassName !== _this3.container.className) {
      _this3.container.className = wrapperClassName;
    }
  };

  this.savePortal = function (c) {
    _this3._component = c;
  };

  this.removeCurrentContainer = function (visible) {
    _this3.container = null;
    _this3._component = null;
    if (!IS_REACT_16) {
      if (visible) {
        _this3.renderComponent({
          afterClose: _this3.removeContainer,
          onClose: function onClose() {},

          visible: false
        });
      } else {
        _this3.removeContainer();
      }
    }
  };
};

exports['default'] = (0, _reactLifecyclesCompat.polyfill)(PortalWrapper);
module.exports = exports['default'];
